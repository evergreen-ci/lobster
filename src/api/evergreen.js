// @flow strict

import type { EvergreenTestLog, EvergreenTaskLogType, EvergreenTaskLog } from '../actions';
import { EVERGREEN_BASE } from '../config';
import { stringToInteralEvergreenTaskLogType } from '../actions';

function taskLogURL(id: string, execution: number, type: EvergreenTaskLogType): string {
  const logType = stringToInteralEvergreenTaskLogType(type) || 'all';
  return `${EVERGREEN_BASE}/task_log_raw/${id}/${execution}?type=${logType}&text=true`;
}

function testLogURL(id: string): string {
  return `${EVERGREEN_BASE}/test_log/${id}?raw=1`;
}

export function fetchEvergreen(log: EvergreenTaskLog | EvergreenTestLog): Promise<Response> {
  const init = { method: 'GET', credentials: 'include' };
  let req;
  if (log.type === 'task') {
    req = new Request(taskLogURL(log.id, log.execution, log.log), init);
  } else if (log.type === 'test') {
    req = new Request(testLogURL(log.id), init);
  }

  return window.fetch(req);
}
