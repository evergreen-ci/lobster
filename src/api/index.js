// @flow strict

export * from './logkeeper';
export * from './evergreen';
