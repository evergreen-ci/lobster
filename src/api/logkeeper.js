// @flow strict

import { LOGKEEPER_BASE } from '../config';

function generateLogkeeperUrl(buildParam: string, testParam: ?string): string {
  if (!buildParam) {
    return '';
  }
  if (!testParam) {
    return LOGKEEPER_BASE + '/build/' + buildParam + '/all?raw=1';
  }
  return LOGKEEPER_BASE + '/build/' + buildParam + '/test/' + testParam + '?raw=1';
}

export function fetchLogkeeper(build: string, test: ?string): Promise<Response> {
  const req = new Request(generateLogkeeperUrl(build, test), { method: 'GET' });
  return window.fetch(req);
}

export function fetchLobster(server: string, url: string): Promise<Response> {
  const init = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      url: url
    })
  };
  const req = new Request(`http://${server}`, init);
  return window.fetch(req);
}
