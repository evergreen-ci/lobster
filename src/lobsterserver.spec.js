import fs from 'fs';
import { tmpdir } from 'os';
import { spawn } from 'child_process';
import path from 'path';
import fetch from 'node-fetch';
import sinon from 'sinon';
import * as api from './api';


function lobster(port = 9000, file = 'simple.log') {
  return fetch(`http://localhost:${port}/api/log`, {
    method: 'POST',
    body: JSON.stringify({ url: file }),
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'Origin': 'http://localhost'
    }
  });
}

function startServer(args) {
  try {
    return spawn('node', [path.dirname(__dirname) + '/server', ...args], {
      stdio: 'inherit'
    });
  } catch (e) {
    console.error(e);
  }
}

describe('lobsterserver', function() {
  beforeEach(() => {
    sinon.replace(window, 'fetch', function(req) {
      console.log(req.url);
      return fetch(req.url);
    });
  });

  let c;
  afterEach(() => {
    sinon.restore();
    if (fs.existsSync('/tmp/lobster.txt')) {
      fs.unlinkSync('/tmp/lobster.txt');
    }
    if (c) {
      c.kill('SIGINT');
      c = undefined;
    }
  });

  e2e('fetch-ok', (done) => {
    c = startServer(['--logs', path.dirname(__dirname) + '/e2e']);
    setTimeout(function() {
      lobster().then((resp) => {
        resp.text().then((body) => {
          expect(resp.status).toBe(200);
          expect(body).toHaveLength(51);
          done();
        });
      }).catch((e) => {
        done.fail(e);
      });
    }, 5000);
  }, 10000);

  e2e('fetch-notexist', (done) => {
    c = startServer(['--logs', path.dirname(__dirname) + '/e2e']);
    setTimeout(function() {
      lobster(undefined, '___notexist.log').then((resp) => {
        resp.text().then((body) => {
          expect(resp.status).toBe(404);
          expect(body).toBe('log not found');
          done();
        });
      }).catch((e) => done.fail(e));
    }, 5000);
  }, 10000);

  e2e('fetch-insecure-path', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer(['--logs', path.dirname(__dirname) + '/e2e']);
    setTimeout(function() {
      lobster(undefined, '___notexist.log').then((resp) => {
        resp.text().then((body) => {
          expect(resp.status).toBe(404);
          expect(body).toBe('log not found');
          done();
        });
      }).catch((e) => done.fail(e));
    }, 5000);
  }, 10000);

  e2e('fetch-logs-disabled', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer([]);
    setTimeout(function() {
      lobster(undefined, 'simple.log').then((resp) => {
        resp.text().then((body) => {
          expect(resp.status).toBe(400);
          expect(body).toBe('');
          done();
        }).catch((e) => done.fail(e));
      }).catch((e) => done.fail(e));
    }, 5000);
  }, 10000);

  e2e('fetch-port', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer(['--port', '8999', '--logs', path.dirname(__dirname) + '/e2e']);
    setTimeout(function() {
      lobster(8999, 'simple.log').then((resp) => {
        resp.text().then((_body) => {
          expect(resp.status).toBe(200);
          done();
        }).catch((e) => done.fail(e));
      }).catch((e) => done.fail(e));
    }, 5000);
  }, 10000);

  e2e('evergreen-test', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer(['--e2e']);
    setTimeout(function() {
      return api.fetchEvergreen({
        type: 'evergreen-test',
        id: 'testid1234'
      }).then((resp) => {
        expect(resp.status).toBe(200);
        return resp.text().then((log) => {
          expect(log).toMatchSnapshot();
          done();
        });
      });
    }, 5000);
  }, 10000);

  e2e('evergreen-task', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer(['--e2e']);
    setTimeout(function() {
      return api.fetchEvergreen({
        type: 'evergreen-task',
        id: 'testid1234',
        execution: 1234
      }).then((resp) => {
        expect(resp.status).toBe(200);
        return resp.text().then((log) => {
          expect(log).toMatchSnapshot();
          done();
        });
      });
    }, 5000);
  }, 10000);

  e2e('logkeeper', (done) => {
    fs.closeSync(fs.openSync(tmpdir() + '/lobster.txt', 'w'));
    c = startServer(['--e2e']);
    setTimeout(function() {
      return api.fetchLogkeeper('build1234', 'test1234')
        .then((resp) => {
          expect(resp.status).toBe(200);
          return resp.text().then((log) => {
            expect(log).toMatchSnapshot();
            done();
          });
        });
    }, 5000);
  }, 10000);
});
