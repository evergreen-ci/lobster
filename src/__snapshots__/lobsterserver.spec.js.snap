// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lobsterserver-default-args e2e-evergreen-task 1`] = `
"testid1234
1234
ALL
line 0
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
something else
enumerate: 0"
`;

exports[`lobsterserver-default-args e2e-evergreen-test 1`] = `
"testid1234
line 0
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
something else
enumerate: 0"
`;

exports[`lobsterserver-default-args e2e-evergreen-test-by-name 1`] = `
"taskid1234
5
testid1234
line 0
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
something else
enumerate: 0"
`;

exports[`lobsterserver-default-args e2e-fetch-ok 1`] = `
"line 0
line 1
line 2
line 3
line 4
somethingelse 5
"
`;

exports[`lobsterserver-default-args e2e-logkeeper 1`] = `
"build1234
test1234
line 0
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
something else
enumerate: 0"
`;

exports[`lobsterserver-default-args e2e-perf-gen 1`] = `
"This log is a performance and rendering stress test for lobster
Chrome has a limit of about 1.67 million lines
If you or selenium can see FIND_THIS_TOKEN at the end of this
file, then you've avoided a performance regression.
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
line 10
FIND_THIS_TOKEN"
`;
